File: spikes/language/ts-bun/src/walk.ts
import { readdirSync, statSync } from "fs";
import { join } from "path";

const dirPath = process.argv[2];
const iterations = parseInt(process.argv[3] || "5", 10);

if (!dirPath) {
  console.error("Usage: bun run src/walk.ts <directory> [iterations]");
  process.exit(1);
}

// Gitignore-aware walk using Bun's Glob
// Bun has a built-in Glob API: new Bun.Glob(pattern).scan(options)

async function walkWithBunGlob(dir: string): Promise<{ files: number; bytes: number }> {
  let fileCount = 0;
  let byteCount = 0;
  const glob = new Bun.Glob("**/*");
  for await (const path of glob.scan({ cwd: dir, dot: false, onlyFiles: true })) {
    fileCount++;
    try {
      const st = statSync(join(dir, path));
      byteCount += st.size;
    } catch {
      // skip inaccessible files
    }
  }
  return { files: fileCount, bytes: byteCount };
}

// Warmup walk
const warmup = await walkWithBunGlob(dirPath);

// Timed iterations
const start = performance.now();
let totalFiles = 0;
let totalBytes = 0;

for (let i = 0; i < iterations; i++) {
  const result = await walkWithBunGlob(dirPath);
  totalFiles += result.files;
  totalBytes += result.bytes;
}

const elapsed = performance.now() - start;
const avgFiles = Math.round(totalFiles / iterations);
const avgBytes = Math.round(totalBytes / iterations);
const avgMs = elapsed / iterations;

console.log(
  JSON.stringify(
    {
      tool: "ts-bun",
      operation: "file_tree_walk",
      directory: dirPath,
      warmup_files: warmup.files,
      iterations,
      total_ms: parseFloat(elapsed.toFixed(2)),
      avg_ms: parseFloat(avgMs.toFixed(3)),
      avg_files: avgFiles,
      avg_bytes: avgBytes,
    },
    null,
    2
  )
);

File: spikes/language/rust/src/walk.rs
use std::env;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: spike-walk <directory> [iterations]");
        std::process::exit(1);
    }
    let dir_path = &args[1];
    let iterations: u32 = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(5);

    // Warmup walk
    let mut warmup_count = 0u64;
    for entry in ignore::WalkBuilder::new(dir_path)
        .hidden(false)
        .git_ignore(true)
        .build()
    {
        if let Ok(e) = entry {
            if e.file_type().is_some_and(|ft| ft.is_file()) {
                warmup_count += 1;
            }
        }
    }

    // Timed iterations
    let start = Instant::now();
    let mut total_files = 0u64;
    let mut total_bytes = 0u64;
    for _ in 0..iterations {
        let mut file_count = 0u64;
        let mut byte_count = 0u64;
        for entry in ignore::WalkBuilder::new(dir_path)
            .hidden(false)
            .git_ignore(true)
            .build()
        {
            if let Ok(e) = entry {
                if e.file_type().is_some_and(|ft| ft.is_file()) {
                    file_count += 1;
                    if let Ok(meta) = e.metadata() {
                        byte_count += meta.len();
                    }
                }
            }
        }
        total_files += file_count;
        total_bytes += byte_count;
    }
    let elapsed = start.elapsed();

    let avg_files = total_files / iterations as u64;
    let avg_bytes = total_bytes / iterations as u64;
    let avg_ms = elapsed.as_secs_f64() * 1000.0 / iterations as f64;
    let total_ms = elapsed.as_secs_f64() * 1000.0;

    println!("{{");
    println!("  \"tool\": \"rust\",");
    println!("  \"operation\": \"file_tree_walk\",");
    println!("  \"directory\": \"{dir_path}\",");
    println!("  \"warmup_files\": {warmup_count},");
    println!("  \"iterations\": {iterations},");
    println!("  \"total_ms\": {total_ms:.2},");
    println!("  \"avg_ms\": {avg_ms:.3},");
    println!("  \"avg_files\": {avg_files},");
    println!("  \"avg_bytes\": {avg_bytes}");
    println!("}}");
}
